pandas/core/generic.py

```diff
@@ -4732,17 +4732,33 @@ def _where(self, cond, other=np.nan, inplace=False, axis=None, level=None,
             cond, _ = cond.align(self, join='right', broadcast_axis=1)
         else:
             if not hasattr(cond, 'shape'):
-                raise ValueError('where requires an ndarray like object for '
-                                 'its condition')
+                cond = np.asanyarray(cond)
             if cond.shape != self.shape:
                 raise ValueError('Array conditional must be same shape as '
                                  'self')
             cond = self._constructor(cond, **self._construct_axes_dict())
 
-        if inplace:
-            cond = -(cond.fillna(True).astype(bool))
+        # If 'inplace' is True, we want to fill with True
+        # before inverting. If 'inplace' is False, we will
+        # fill with False and do nothing else.
+        #
+        # Conveniently, 'inplace' matches the boolean with
+        # which we want to fill.
+        cond = cond.fillna(bool(inplace))
+
+        msg = "Boolean array expected for the condition, not {dtype}"
+
+        if not isinstance(cond, pd.DataFrame):
+            # This is a single-dimensional object.
+            if not is_bool_dtype(cond):
+                raise ValueError(msg.format(dtype=cond.dtype))
         else:
-            cond = cond.fillna(False).astype(bool)
+            for dt in cond.dtypes:
+                if not is_bool_dtype(dt):
+                    raise ValueError(msg.format(dtype=dt))
+
+        cond = cond.astype(bool, copy=False)
```

Comment:
Wait, are you saying the original code on `master` was simpler to understand!?  I fail to see why.